WordPress/gutenberg
# REST API: Use posts endpoint for reusable blocks

Related: #7453
Related: https://core.trac.wordpress.org/ticket/45098

This pull request seeks to refactor reusable blocks to avoid the need for (and thus enable removal of) the blocks API controller, in favor of the existing posts controller. **Edit 2018-10-20:** The controller class remains, solely for its custom `check_read_permission` implementation to avoid public visibility of published reusable blocks.

In implementing these changes, the custom controller was found to be used to:

1. Return a limited subset of post endpoint properties (`id`, `title`, `content`)
2. Flatten `title` and `content` fields (`raw` string used, no object of `raw`/`rendered`)
3. Force delete on trash
4. Force publish on save

My assessment of this is:

- The first two points introduce a needless and unexpected inconsistency from the posts response
- The last two points should be enforced at a post-type level if desirable
   - Are they desirable?
- A subtlety of the second point is that now a user needs to have edit capabilities for the `wp_block` type to be able to view it in the editor, or to export the block. This is because the raw content (`post.content.raw`) is only included when the `context=edit` parameter is passed
   - This is perhaps a main deterrent to proceeding here, though is symptomatic of a larger issue around REST permissions where the prior implementation was an ad hoc workaround (effect of bumping permissions for otherwise unqualified users).

Post type supports `title` and `editor` were required to be added for the `title` and `content` fields to be included in the REST response [[1]](https://github.com/WordPress/WordPress/blob/da7a80d67fea29c2badfc538bfc01c8a585f0cbe/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L1978-L2001) [[2]](https://github.com/WordPress/WordPress/blob/da7a80d67fea29c2badfc538bfc01c8a585f0cbe/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L2003-L2032). See also related discussion in #9964 (the changes here enable a Gutenberg editor for the Reusable Blocks management experience, thereby effectively closing #9964).

**Open questions:**

Should there be / how would there be a deprecation migration period for the controller removal?

**Testing Instructions:**

Verify no regressions in behavior of reusable blocks:

- Creating / updating
- Inserting
- Managing all
- Single block vs. template
- Import / export
